docs/design/readme.md

```diff
@@ -0,0 +1,1704 @@
+    // state is not lost by creating a dummy client instance whose only purpose is to keep the counter alive.
+    template <typename Service>
+    [[nodiscard]] std::expected<Client<Service>, Error> makeClient(const std::uint16_t service_id,
+                                                                                      const std::uint16_t server_node_id);
+
+    // If such a server already exists, an error will be returned (error type TBD).
+    template <typename Service>
+    [[nodiscard]] std::expected<UniquePtr<Server<Service>>, Error> getServer(const std::uint16_t service_id);
+
+    [[nodiscard]] transport::ITransport& getTransport() noexcept;
+    [[nodiscard]] const transport::ITransport& getTransport() const noexcept;
+};
+
+/// The hidden per-port-unique implementation classes (not visible to the user, managed automatically).
+class PublisherImpl;
+class SubscriberImpl : public IRunnable;
+class ClientImpl : public IRunnable;
+```
+
+### Publisher
+
+```c++
+// One of the purposes of this class is to reduce the amount of generic code to the minimum.
+// Only those parts that are not message-type-invariant are implemented in the concrete derived class.
+// Items are movable but not copyable.
+class PublisherBase
+{
+public:
+    using Error = std::variant<AnonymousError, ArgumentError, MemoryError, CapacityError, PlatformError>;
+
+    [[nodiscard]] transport::IMessageTxSession& getSession() noexcept;
+    [[nodiscard]] const transport::IMessageTxSession& getSession() const noexcept;
+
+    /// The default priority is nominal.
+    void setPriority(const Priority prio) noexcept;
+    [[nodiscard]] Priority getPriority() noexcept;
+
+    /// Publish a pre-serialized message. For the overload that accepts the typed message object see the derived class.
+    /// The transport frames constructed from this message will be sent to the wire before the expiration of the
+    /// deadline or discarded.
+    [[nodiscard]] std::expected<void, Error> publish(const TimePoint deadline,
+                                                     const std::span<const std::byte> data)
+    {
+        return impl_->publish(deadline, prio_, data);
+    }
+
+    // TODO: helper methods setTimeout()/getTimeout() may be added to support simpler publish overloads that
+    // do not take the deadline but rather compute it automatically from the current time and the timeout.
+
+    // The type is movable and non-copyable.
+    PublisherBase(const PublisherBase&)            = delete;
+    PublisherBase(PublisherBase&&)                 = default;
+    PublisherBase& operator=(const PublisherBase&) = delete;
+    PublisherBase& operator=(PublisherBase&&)      = default;
+    virtual ~PublisherBase() noexcept              = default;
+
+protected:
+    explicit PublisherBase(const SharedPtr<PublisherImpl>& impl);
+
+private:
+    SharedPtr<PublisherImpl> impl_;
```

Comment:
Let's talk about `SharedPtr` here... Why does PublisherImpl have to be reference counted? Besides the total lack of control over an object's lifecycle and the enabling of very unsafe multi-threaded code, sharing pointers indiscriminately is normally just lazy and rarely required once the use case is properly analysed. If the requirement is reuse of dynamic memory then a locking semantic is easier to audit. For example, using something like:

```
std::expected<std::unique_ptr<IDataObject>, Error> record = obj.lockNextRecord();
```
provides the user with clear semantics that the caller has exclusive access to the data object, if it is returned. Errors could include "already locked" which would prevent concurrent, mutable access to the data object. To allow multiple readers something like this could be provided:

```
std::expected<std::unique_ptr<const IDataObject>, Error> record = obj.peekNextRecord();
```

The data objects themselves should be created and deleted at well-defined times in the program. This prevents random fluctuations in WCET based on what order shared pointers were acquired and released and by which threads. Instead, given the run-to-completion model libcyphal will enable, we should define when objects will be deleted and check for outstanding locks on this memory as part of this phase. For example: data object deletion could be performed as part of unwinding an IRunnable callstack and any data objects still locked could result in errors. This means any memory leaks would be found within the context of a single run cycle.

Reply:
Publishers operating on the same port share the transfer-ID counter and the transport session. Subscribers operating on the same port share the transport session; additionally, each received message is cloned (in the future cloning can be replaced with some form of sharing to reduce data copying) between the user-visible subscription objects. Clients operating on the same service and same server node share the transfer-ID counter, the transport sessions, and the registry of pending requests/responses.

The user code is not responsible for managing the lifecycle of the hidden shared implementation object because that would introduce coupling between different parts of the user's application, which may be operating on the same port (e.g., multiple components may need to receive data from the same subject or invoke the same service). The lifetime of the hidden object begins when the first object on a port is created and ends when the last object on said port is deleted. Thread safety is a valid concern, but it is not explicitly covered in this document, which is my omission.